beacon_node/beacon_chain/src/eth1_chain.rs

```diff
@@ -89,22 +147,977 @@ impl<T: EthSpec> Eth1ChainBackend<T> for InteropEth1ChainBackend<T> {
         })
     }
 
-    fn queued_deposits(&self, _: &BeaconState<T>) -> Result<Vec<Deposit>> {
+    fn queued_deposits(&self, _: &BeaconState<T>, _: &ChainSpec) -> Result<Vec<Deposit>, Error> {
         Ok(vec![])
     }
 }
 
-impl<T: EthSpec> Default for InteropEth1ChainBackend<T> {
+impl<T: EthSpec> Default for DummyEth1ChainBackend<T> {
     fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+/// Maintains a cache of eth1 blocks and deposits and provides functions to allow block producers
+/// to include new deposits and vote on `Eth1Data`.
+///
+/// The `core` connects to some external eth1 client (e.g., Parity/Geth) and polls it for
+/// information.
+#[derive(Clone)]
+pub struct CachingEth1Backend<T: EthSpec, S> {
+    pub core: HttpService,
+    store: Arc<S>,
+    log: Logger,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: EthSpec, S: Store> CachingEth1Backend<T, S> {
+    /// Instantiates `self` with empty caches.
+    ///
+    /// Does not connect to the eth1 node or start any tasks to keep the cache updated.
+    pub fn new(config: Eth1Config, log: Logger, store: Arc<S>) -> Self {
+        Self {
+            core: HttpService::new(config, log.clone()),
+            store,
+            log,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Starts the routine which connects to the external eth1 node and updates the caches.
+    pub fn start(&self, exit: Exit) -> impl Future<Item = (), Error = ()> {
+        self.core.auto_update(exit)
+    }
+
+    /// Instantiates `self` from an existing service.
+    pub fn from_service(service: HttpService, store: Arc<S>) -> Self {
         Self {
+            log: service.log.clone(),
+            core: service,
+            store,
             _phantom: PhantomData,
         }
     }
 }
 
+impl<T: EthSpec, S: Store> Eth1ChainBackend<T> for CachingEth1Backend<T, S> {
+    fn eth1_data(&self, state: &BeaconState<T>, spec: &ChainSpec) -> Result<Eth1Data, Error> {
```

Comment:
I've done my best to grok this, and I'm fairly sure it matches the spec: https://github.com/ethereum/eth2.0-specs/blob/v0.9.1/specs/validator/0_beacon-chain-validator.md#eth1-data

Reply:
phew!